api/hasura/actions/create_user.ts

```diff
@@ -0,0 +1,100 @@
+import type { VercelRequest, VercelResponse } from '@vercel/node';
+
+import { gql } from '../../../api-lib/Gql';
+import { verifyHasuraAdminMiddleware } from '../../../api-lib/validate';
+import { profiles_constraint } from '../../../src/lib/gql/zeusHasuraAdmin';
+import { createUserSchemaInput } from '../../../src/lib/zod';
+
+async function handler(request: VercelRequest, response: VercelResponse) {
+  // Input Validation
+  const result = await createUserSchemaInput.safeParseAsync(request.body.input);
+  // a `switch` gets typescript to behave soundly
+  // an `if` statement doesn't recognize the result
+  // disambiguation for some reason
+  switch (result.success) {
+    case false:
+      return response.status(422).json({
+        extensions: result.error.issues,
+        message: 'invalid input',
+        code: 422,
+      });
+  }
+
+  // External Constraint Validation
+  // I'd prefer to add this uniqueness constraint into the database
+  const { object: createUserParams, circle_id } = result.data;
+
+  const { users: existingUsers } = await gql.q('query')({
```

Comment:
eventually we will want to have a nice library for all these things, although if it is unlikely to be reused it might as well be kept in the action

Reply:
once we get a uniqueness contraint on the users table this particular query should be unnecessary